source/util/move_to_front.h

```diff
@@ -0,0 +1,325 @@
+  // Before the call: 4 8 2 1 7
+  // RankFromId(8) returns 1
+  // After the call: 8 4 2 1 7
+  // If id is not in the list, and is equal to next_id_, then a new value is
+  // inserted at the front. The function returns 0 in this case.
+  // RankFromId(9) returns 0
+  // After the call: 9 8 4 2 1 7
+  // Calling the function with a value which is not in the list and is not
+  // equal to next_id_ will result in assertion.
+  size_t RankFromId(uint32_t id);
+
+  // Returns id corresponding to a 1-indexed rank in the move-to-front sequence
+  // and moves the id to the front. Example:
+  // Before the call: 4 8 2 1 7
+  // IdFromRank(1) returns 8
+  // After the call: 8 4 2 1 7
+  // If rank is 0, then a new id equal to next_id_ is inserted.
+  // IdFromRank(0) returns 9
+  // After the call:  9 8 4 2 1 7
+  uint32_t IdFromRank(size_t rank);
+
+  // Permanently removes the id from the move-to-front sequence.
+  // The implementation has artificially restricted functionality which only
+  // allows to add new ids sequentially. So once the id is removed using this
+  // function it cannot be reinserted.
+  void DeprecateId(uint32_t id) {
+    auto it = id_to_node_.find(id);
+    assert(it != id_to_node_.end());
+    RemoveNode(it->second);
+    // The iterator should still be valid, even if RemoveNode has modified
+    // id_to_node_. But just in case erase by id, not by iterator.
+    id_to_node_.erase(id);
+  }
+
+  // Returns the number of elements in the move-to-front sequence.
+  size_t GetSize() const {
+    return SizeOf(root_);
+  }
+
+  // The methods below are for testing only.
+
+  // Inserts the value in the internal tree data structure. For testing only.
+  void TestInsert(uint32_t val) {
+    InsertNode(CreateNode(val, val));
+  }
+
+  // Removes the value from the internal tree data structure. For testing only.
+  void TestRemove(uint32_t val) {
+    const auto it = id_to_node_.find(val);
+    assert (it != id_to_node_.end());
+    RemoveNode(it->second);
+  }
+
+  // Prints the internal tree data structure to |out|. For testing only.
+  void PrintTree(std::ostream& out, bool print_timestamp = false) const {
+    if (root_)
+      PrintTreeInternal(out, root_, 1, print_timestamp);
+  }
+
+ private:
+  // Internal tree data structure uses handles instead of pointers. Leaves and
+  // root parent reference a singleton under handle 0. Although dereferencing
+  // a null pointer is not possible, inappropriate access to handle 0 would
+  // cause an assertion. Handles are not garbage collected if id was deprecated
+  // with DeprecateId(). But handles are recycled when a node is repositioned.
+
+  // Internal tree data structure node.
+  struct Node {
+    // Fields are in order from biggest bit width to smallest.
+    // SPIR-V id.
+    uint32_t id = 0;
+    // Timestamp from a logical clock which updates every time the element is
+    // accessed through IdFromRank or RankFromId.
+    uint32_t timestamp = 0;
```

Comment:
Feedback unaddressed from before.  What happens when timestamp rolls over?
I'd be happier with this as a 64-bit type.